### PR TITLE
fix `treat_as_floating_point_v<_DurationType>`

### LINKED ISSUE #2636: chrono: Questionable usage of treat_as_floating_point_v<_DurationType>
In header `<chrono>`, line https://github.com/microsoft/STL/blob/27877181dc50fc5f0dc9d679703437eb105e2b9f/stl/inc/chrono#L3392, within function `_Apply_duration_fields`, an if condition contains the expression 
`treat_as_floating_point_v<_DurationType>` where `_DurationType` represents a `std::chrono::duration` specialization. But `std::chrono::treat_as_floating_point_v` ([time.traits.is.fp]) doesn't expect a duration, but performs the test directly on the provided type (usually the representation type of a duration), so this part of the if condition will always evaluate to `false`, unless a programmer would specialize a duration with a user-provided period or representation type for `std::chrono::treat_as_floating_point`, which seems unlikely. It seems to me that the mentioned code location intended to use `treat_as_floating_point_v<typename _DurationType::rep>` instead.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2636

### CODE REVIEW

File: stl/inc/chrono [hunk 1]
```diff
@@ -3389,12 +3389,15 @@ namespace chrono {
                     if (_Subsecond) {
                         using _CastedType = duration<int64_t, typename _DurationType::period>;
                         const _SubsecondType _Sub{*_Subsecond};
-                        if (treat_as_floating_point_v<_DurationType> //
+#pragma warning(push)
+#pragma warning(disable : 4127) // conditional expression is constant
+                        if (treat_as_floating_point_v<typename _DurationType::rep> //
```
Comment (StephanTLavavej) [inline stl/inc/chrono:3394]: While this is correct, we could avoid the need for push-disable-pop by refactoring this with `if constexpr` and De Morgan's law.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! (I especially appreciate the added test coverage.)

To avoid the need to suppress compiler warnings, I went ahead and pushed what looks like a major rearrangement, but is actually a few simple transformations:

* Use De Morgan's law to flip the condition.
* After an early `return`, we don't need `else`.
* Separate `if (compile_time_condition && runtime_condition)` into nested `if constexpr` and `if`.
* Reduce the scope of `using _CastedType` to where it's needed.

I believe that the resulting code reads naturally: "At compile-time, if we're not treating the rep as floating-point, then we're treating it as integral (which might lose precision). We'll need `_CastedType`. Perform a runtime check for precision loss - if `duration_cast` returns a non-equal value, we've lost precision and must immediately `return false`. Otherwise, we're good to go for updating `_Result`."

For the treat-as-floating-point case, there is an extremely minor throughput advantage to `if constexpr` (don't need to instantiate `duration_cast<_CastedType>`) although I was originally motivated by the compiler warning suppression.

Hope this is okay - it's a bit larger than the changes I usually push without asking.

**Comment by fsb4000** [comment]
Good. 
Actually Visual C++ compiler said me the same: "C4127: conditional expression is constant. Consider using 'if constexpr' statement instead"
But I looked at the code and I found that we have 3 other places with:
```
#pragma warning(push)
#pragma warning(disable : 4127) // conditional expression is constant
```
So this is why I did "pragma push"...
I should have listened to the compiler :)

**Comment by StephanTLavavej** [comment]
Ah, thanks - I had forgotten we still had some of those. I'll go through and clean up (in a separate PR) as many of the 4127 suppressions as possible/convenient (most of them date back to when we couldn't use `if constexpr` unconditionally). The ones in `<span>` are probably more trouble to change than they're worth.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks again for fixing this bug and testing the affected parsing scenario! :heart_eyes_cat: :lady_beetle: :tada:

**Pull request merged**